### PR TITLE
[3979] Add Project related REST APIs in Sirius Web

### LINKED ISSUE #3979: Add support for a REST API
See the [Systems Modeling Application Programming Interface (API) and Services](https://www.omg.org/spec/SystemsModelingAPI/1.0/Beta1/PDF)  section "8.1 REST/HTTP PSM". We will consider at least both the `ProjectService` and `ElementNavigationService`.

Additional resources:

- https://www.omg.org/spec/SystemsModelingAPI/20230201/OpenAPI.json
- https://www.omg.org/spec/SystemsModelingAPI/20230201/Schemas.json

### PR DESCRIPTION
**Pull request by AxelRICHARD**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3979

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [x] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        var createProjectInput = new CreateProjectInput(UUID.randomUUID(), name, List.of());
+        var newProjectPayload = this.projectApplicationService.createProject(createProjectInput);
+
+        if (newProjectPayload instanceof CreateProjectSuccessPayload createProjectSuccessPayload) {
+            var projectDTO = createProjectSuccessPayload.project();
+            var restProject = new RestProject(projectDTO.id(), null, List.of(), null, null, projectDTO.name(), List.of());
+            return new ResponseEntity<>(restProject, HttpStatus.CREATED);
+        }
+        // The specification does not handle other HttpStatus than HttpStatus.CREATED for this endpoint
+        return null;
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java:88]: I should test that but if you know, what happens in this case? Do we have at least a proper HTTP response?

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/data/versioning/dto/IRestCommitReference.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import org.eclipse.sirius.web.application.dto.IRestRecord;
+
+/**
+ * Interface for the REST CommitReference DTO.
+ *
+ * @author arichard
+ */
+public interface IRestCommitReference extends IRestRecord {
+
+    DateFormat created();
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/data/versioning/dto/IRestCommitReference.java]: Please prefer `Instant`, `DateFormat` is not a date, it's a java.text utility class to help formatting dates. We can see how the instant will be formatted later.

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/data/versioning/dto/IRestCommitReference.java (hunk 2 above)
Comment (AxelRICHARD) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/data/versioning/dto/IRestCommitReference.java]: Yes, sorry I missed that.

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java (hunk 1 above)
Comment (AxelRICHARD) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java:88]: Yes, but a `200 OK` one. I don't know If we should return another kind of HTTP response, even if the specification does not handle the case.

Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java:88]: If it fallback on a default value that's ok for me. The specification should probably indicate what happens if when can't perform the action, the obvious one would be BAD_REQUEST but who knows.

:+1:

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    @PutMapping(path = "/{projectId}")
+    public ResponseEntity<RestProject> updateProject(@PathVariable UUID projectId, @RequestParam Optional<String> name, @RequestParam Optional<String> description, @RequestParam Optional<RestBranch> branch) {
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java:92]: It's a bit odd that the API would expect as an input a RestBranch as a request param. How would that be serialized? Would the end user send the whole object?

I don't care to move forward and merge the PR, it's just odd but it's a first version.

**Review by sbegaudeau — APPROVED** [review]

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java (hunk 3 above)
Comment (AxelRICHARD) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java:92]: You think it should be a `@RequestBody` instead?

Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectRestController.java:92]: Well the specification is definitely not clear on this topic and [the live instance](http://sysml2.intercax.com:9000/docs/#put-/projects/-projectId-) to test things seems to use a different approach with the following body to update a project:

```
{
  "@type": "Project",
  "defaultBranch": {
    "@id": "3fa85f64-5717-4562-b3fc-2c963f66afa6"
  },
  "description": "string",
  "name": "string"
}
```

What part of this is mandatory? what is the schema to use? If I understand things clearly from [their example](https://github.com/Systems-Modeling/SysML-v2-API-Services/blob/master/app/controllers/ProjectController.java#L61) and from this sample it seems that one has to send the full body of the project to update a project. The code of the play based demo seems very odd, I really don't understand the point of something like that:

```
        Project requestedObject = Json.fromJson(requestBodyJson, metamodelProvider.getImplementationClass(Project.class));
        if (requestedObject.getId() != null && !id.equals(requestedObject.getId())) {
            return Results.badRequest();
        }
        requestedObject.setId(id);
```

Given that the identifier is given as a path parameter. Sending back the whole project seems painful to just update a field. By the way, have you seen [this](https://github.com/Systems-Modeling/SysML-v2-API-Services/tree/master/conf/json/schema/metamodel)?

Anyway, let's merge a first version and we will update it later :+1:

**Pull request merged**